pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java

```diff
@@ -916,30 +916,32 @@ private void doPendingBatchReceiveTask(Timeout timeout) {
     protected void triggerListener() {
         // Trigger the notification on the message listener in a separate thread to avoid blocking the networking
         // thread while the message processing happens
-        try {
-            // Control executor to call MessageListener one by one.
-            if (executorQueueSize.get() < 1) {
-                final Message<T> msg = internalReceive(0, TimeUnit.MILLISECONDS);
-                if (msg != null) {
-                    executorQueueSize.incrementAndGet();
-                    if (SubscriptionType.Key_Shared == conf.getSubscriptionType()) {
-                        executorProvider.getExecutor(peekMessageKey(msg)).execute(() ->
-                                callMessageListener(msg));
-                    } else {
-                        getExternalExecutor(msg).execute(() -> {
-                            callMessageListener(msg);
-                        });
+        internalPinnedExecutor.execute(() -> {
+            try {
+                // Control executor to call MessageListener one by one.
+                if (executorQueueSize.get() < 1) {
+                    final Message<T> msg = internalReceive(0, TimeUnit.MILLISECONDS);
+                    if (msg != null) {
+                        executorQueueSize.incrementAndGet();
```

Comment:
Here is the race condition. Lines 922 and 925. The non-synchronized get and subsequent update would explain messages processed out of order. By putting this on the same thread, `internalPinnedExecutor`, the race is no longer possible. However, I wonder if we want the extra thread switching here and if we want the `internalReceive` method called on the `internalPinnedExecutor` for _all_ method calls. It seems like we should have instead solved this data race.

Reply:
After looking at the client a bit more, we certainly follow this paradigm frequently. It is possible that this `internalPinnedExecutor` might become a bit of a bottle neck. However, I think we can keep the design as is for now. I opened https://github.com/apache/pulsar/pull/13273 to clean up this code block a bit.